File: packages/governance-py/src/aumos_governance/budget/manager.py
# SPDX-License-Identifier: BSL-1.1
# Copyright (c) 2026 MuVeraAI Corporation
from __future__ import annotations

from pydantic import BaseModel

from aumos_governance.budget.policy import apply_rollover, should_reset
from aumos_governance.budget.tracker import CategoryTracker, SpendingTransaction
from aumos_governance.config import BudgetConfig
from aumos_governance.errors import BudgetExceededError, BudgetNotFoundError, InvalidPeriodError
from aumos_governance.types import BUDGET_PERIOD_VALUES


class BudgetCheckResult(BaseModel, frozen=True):
    """
    Result of a budget availability check.

    Attributes:
        allowed: True if the requested amount is within the available budget.
        category: The budget category evaluated.
        requested: The amount that was checked.
        available: The remaining budget at the time of the check.
        limit: The total configured limit for the category.
        spent: The amount already spent in the current period.
        reason: Human-readable explanation of the outcome.
    """

    allowed: bool
    category: str
    requested: float
    available: float
    limit: float
    spent: float
    reason: str


class BudgetManager:
    """
    Manages static spending budgets for one or more categories.

    Budget allocations are ALWAYS static. There is no adaptive limit
    adjustment, spending prediction, or ML-based optimization.

    All data is stored in-memory. A new BudgetManager starts empty.

    Example::

        manager = BudgetManager(BudgetConfig(allow_overdraft=False))
        manager.create_budget("llm-calls", limit=100.0, period="monthly")
        manager.record_spending("llm-calls", 5.0, description="GPT-4o request")
        result = manager.check_budget("llm-calls", 10.0)
        assert result.allowed is True
    """

    def __init__(self, config: BudgetConfig | None = None) -> None:
        self._config = config or BudgetConfig()
        self._tracker = CategoryTracker()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def create_budget(
        self,
        category: str,
        limit: float,
        period: str = "monthly",
    ) -> None:
        """
        Create a new budget envelope for a spending category.

        Budgets are static: the limit you set here is the limit for every
        period until you replace the budget. There is no automatic adjustment.

        Args:
            category: Unique name for this budget (e.g. ``'llm-calls'``).
            limit: Maximum spend allowed per period. Must be >= 0.
            period: Reset period — one of ``'daily'``, ``'weekly'``,
                ``'monthly'``, ``'yearly'``, or ``'lifetime'``.

        Raises:
            ValueError: If ``category`` already exists, or if ``limit`` < 0.
            InvalidPeriodError: If ``period`` is not a recognised value.
        """
        if limit < 0:
            raise ValueError(f"Budget limit must be >= 0; got {limit}.")
        if period not in BUDGET_PERIOD_VALUES:
            raise InvalidPeriodError(period)
        self._tracker.create(category=category, limit=limit, period=period)

    def record_spending(
        self,
        category: str,
        amount: float,
        description: str | None = None,
    ) -> SpendingTransaction:
        """
        Record a spending transaction against a budget category.

        If :attr:`~BudgetConfig.allow_overdraft` is False and the
        transaction would exceed the budget, a
        :class:`~aumos_governance.errors.BudgetExceededError` is raised and
        **no state is mutated**.

        If the category's period has elapsed, the budget is automatically
        reset before recording the transaction.

        Args:
            category: The budget category to charge.
            amount: Amount to record. Must be positive.
            description: Optional description stored with the transaction.

        Returns:
            The created :class:`~aumos_governance.budget.tracker.SpendingTransaction`.

        Raises:
            BudgetNotFoundError: If ``category`` does not exist.
            BudgetExceededError: If the transaction would exceed the limit
                and overdraft is not allowed.
            ValueError: If ``amount`` is not positive.
        """
        envelope = self._tracker.get(category)
        if envelope is None:
            raise BudgetNotFoundError(category)

        self._maybe_reset(category)
        # Re-fetch after potential reset.
        envelope = self._tracker.get(category)
        assert envelope is not None  # noqa: S101 — guaranteed above

        if not self._config.allow_overdraft:
            projected = envelope.spent + amount
            if projected > envelope.effective_limit:
                raise BudgetExceededError(
                    category=category,
                    requested=amount,
                    available=envelope.remaining,
                )

        return self._tracker.record(category=category, amount=amount, description=description)

    def check_budget(
        self,
        category: str,
        amount: float,
    ) -> BudgetCheckResult:
        """
        Check whether a spending amount is within the available budget.

        This is a read-only operation — it does not record any spending
        or modify any state.

        Args:
            category: The budget category to check.
            amount: The amount to check against the remaining budget.

        Returns:
            A :class:`BudgetCheckResult` describing the outcome.

        Raises:
            BudgetNotFoundError: If ``category`` does not exist.
        """
        envelope = self._tracker.get(category)
        if envelope is None:
            raise BudgetNotFoundError(category)

        available = envelope.remaining
        allowed = amount <= available

        if allowed:
            reason = (
                f"Category '{category}': {amount:.4f} requested, "
                f"{available:.4f} available ({envelope.spent:.4f} of "
                f"{envelope.effective_limit:.4f} spent)."
            )
        else:
            reason = (
                f"Category '{category}': {amount:.4f} requested but only "
                f"{available:.4f} remains ({envelope.spent:.4f} of "
                f"{envelope.effective_limit:.4f} spent)."
            )

        return BudgetCheckResult(
            allowed=allowed,
            category=category,
            requested=amount,
            available=available,
            limit=envelope.effective_limit,
            spent=envelope.spent,
            reason=reason,
        )

    def get_utilization(self, category: str) -> float:
        """
        Return the fraction of the budget consumed for ``category``.

        Returns:
            A float between 0.0 and 1.0 (may exceed 1.0 if overdraft occurred).

        Raises:
            BudgetNotFoundError: If ``category`` does not exist.
        """
        envelope = self._tracker.get(category)
        if envelope is None:
            raise BudgetNotFoundError(category)
        return envelope.utilization

    def list_categories(self) -> list[str]:
        """Return all registered budget category names."""
        return self._tracker.all_categories()

    def summary(self) -> list[dict[str, object]]:
        """
        Return a summary snapshot of all budget envelopes.

        Returns:
            List of dicts, one per category, with fields:
            ``category``, ``limit``, ``effective_limit``, ``spent``,
            ``remaining``, ``utilization``, ``period``, ``last_reset``,
            ``transaction_count``.
        """
        return self._tracker.snapshot()

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _maybe_reset(self, category: str) -> None:
        """Reset the budget for ``category`` if the period has elapsed."""
        envelope = self._tracker.get(category)
        if envelope is None:
            return
        if should_reset(period=envelope.period, last_reset=envelope.last_reset):
            new_limit = apply_rollover(
                spent=envelope.spent,
                limit=envelope.limit,
                rollover_on_reset=self._config.rollover_on_reset,
            )
            self._tracker.reset(category=category, new_effective_limit=new_limit)

File: packages/governance-ts/src/budget/manager.ts
// SPDX-License-Identifier: BSL-1.1
// Copyright (c) 2026 MuVeraAI Corporation

import type { SpendingEnvelope, BudgetPeriod, BudgetCheckResult, BudgetUtilization } from '../types.js';
import type { BudgetConfig } from '../config.js';
import { parseBudgetConfig } from '../config.js';
import { SpendingTracker, computeNextResetAt } from './tracker.js';
import { applyRolloverIfDue } from './policy.js';

/**
 * BudgetManager tracks per-category spending envelopes and enforces static
 * spending limits.
 *
 * Budget allocation is static — limits are set at construction or via
 * createBudget().  There is no mechanism for dynamic limit adjustment based
 * on spend patterns or ML signals.
 *
 * Storage is an in-memory Map keyed by category name.  Each entry is a
 * mutable SpendingEnvelope; SpendingTracker holds the transaction log.
 *
 * Public API (Fire Line — do NOT add methods beyond these three core ones):
 *   createBudget()    — create or replace a spending envelope
 *   recordSpending()  — record a settled transaction against a category
 *   checkBudget()     — check whether a requested amount fits within limits
 *
 * Additional read-only helper:
 *   getUtilization()  — snapshot of current utilisation for a category
 */
export class BudgetManager {
  readonly #config: BudgetConfig;
  /** Primary storage keyed by category name. */
  readonly #envelopes: Map<string, SpendingEnvelope> = new Map();
  readonly #tracker: SpendingTracker = new SpendingTracker();

  constructor(config: unknown = {}) {
    this.#config = parseBudgetConfig(config);

    // Seed envelopes declared in configuration.
    if (this.#config.envelopes !== undefined) {
      for (const preset of this.#config.envelopes) {
        this.createBudget(preset.category, preset.limit, preset.period);
      }
    }
  }

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  /**
   * Fetches an envelope, applying period rollover first so that limit checks
   * always operate on the current period's spend figures.
   */
  #getActiveEnvelope(category: string): SpendingEnvelope | undefined {
    const envelope = this.#envelopes.get(category);
    if (envelope === undefined) {
      return undefined;
    }
    applyRolloverIfDue(envelope, new Date());
    return envelope;
  }

  /**
   * Computes the aggregate amount spent today across all categories.
   * Used when a dailyLimitUsd is configured.
   */
  #computeTotalDailySpend(): number {
    let total = 0;
    for (const envelope of this.#envelopes.values()) {
      if (envelope.period === 'daily' || envelope.period === 'total') {
        total += envelope.spent;
      }
    }
    return total;
  }

  // ---------------------------------------------------------------------------
  // Public API
  // ---------------------------------------------------------------------------

  /**
   * Creates a new spending envelope for the given category.
   *
   * If an envelope already exists for this category it is replaced.
   * The new envelope starts with zero spend.
   *
   * @param category - Unique label for this budget category.
   * @param limit    - Maximum allowable spend in the period.
   * @param period   - Period over which spend accumulates before resetting.
   * @returns The newly created SpendingEnvelope.
   */
  createBudget(category: string, limit: number, period: BudgetPeriod): SpendingEnvelope {
    if (category.trim().length === 0) {
      throw new RangeError('category must be a non-empty string.');
    }
    if (limit <= 0) {
      throw new RangeError('limit must be a positive number.');
    }

    const now = new Date();
    const envelope: SpendingEnvelope = {
      id: crypto.randomUUID(),
      category,
      limit,
      period,
      spent: 0,
      committed: 0,
      resetAt: computeNextResetAt(period, now),
      createdAt: now.toISOString(),
    };

    this.#envelopes.set(category, envelope);
    return envelope;
  }

  /**
   * Records a settled spending transaction against the named category.
   *
   * Throws RangeError when:
   * - The category does not have a registered envelope.
   * - The amount is not a positive finite number.
   *
   * Note: this method records unconditionally — callers should call
   * checkBudget() first and only call recordSpending() when the result
   * is `permitted: true`.
   *
   * @param category    - The budget category to debit.
   * @param amount      - Amount to record (must be positive).
   * @param description - Optional human-readable description.
   */
  recordSpending(category: string, amount: number, description?: string): void {
    if (!isFinite(amount) || amount <= 0) {
      throw new RangeError('amount must be a positive finite number.');
    }

    const envelope = this.#getActiveEnvelope(category);
    if (envelope === undefined) {
      throw new RangeError(
        `No spending envelope found for category "${category}". ` +
          `Call createBudget() first.`,
      );
    }

    this.#tracker.recordTransaction(envelope, amount, description);
  }

  /**
   * Checks whether a requested spend amount is within budget for the
   * given category.
   *
   * Returns `permitted: false` when:
   * - No envelope exists for the category.
   * - The amount would exceed the envelope limit.
   * - A global `dailyLimitUsd` is configured and would be exceeded.
   *
   * This method never mutates state — it is safe to call repeatedly without
   * side effects.
   *
   * @param category  - The budget category to check.
   * @param amount    - The prospective spend amount.
   * @returns BudgetCheckResult with full context for audit logging.
   */
  checkBudget(category: string, amount: number): BudgetCheckResult {
    const envelope = this.#getActiveEnvelope(category);

    if (envelope === undefined) {
      return {
        permitted: false,
        available: 0,
        requested: amount,
        limit: 0,
        spent: 0,
        reason: `No spending envelope registered for category "${category}".`,
      };
    }

    const available = envelope.limit - envelope.spent - envelope.committed;
    const withinEnvelope = amount <= available;

    if (!withinEnvelope) {
      return {
        permitted: false,
        available: Math.max(0, available),
        requested: amount,
        limit: envelope.limit,
        spent: envelope.spent,
        reason:
          `Requested amount ${amount} exceeds available budget ${Math.max(0, available)} ` +
          `for category "${category}" (limit: ${envelope.limit}, spent: ${envelope.spent}).`,
      };
    }

    // Check global daily limit if configured.
    if (this.#config.dailyLimitUsd !== undefined) {
      const totalDailySpend = this.#computeTotalDailySpend();
      const dailyAvailable = this.#config.dailyLimitUsd - totalDailySpend;

      if (amount > dailyAvailable) {
        return {
          permitted: false,
          available: Math.max(0, dailyAvailable),
          requested: amount,
          limit: envelope.limit,
          spent: envelope.spent,
          reason:
            `Requested amount ${amount} would exceed the global daily limit. ` +
            `Daily available: ${Math.max(0, dailyAvailable)}.`,
        };
      }
    }

    return {
      permitted: true,
      available: Math.max(0, available),
      requested: amount,
      limit: envelope.limit,
      spent: envelope.spent,
    };
  }

  /**
   * Returns a read-only utilisation snapshot for the named category.
   *
   * Returns undefined when no envelope exists for the category.
   */
  getUtilization(category: string): BudgetUtilization | undefined {
    const envelope = this.#getActiveEnvelope(category);
    if (envelope === undefined) {
      return undefined;
    }

    const available = Math.max(0, envelope.limit - envelope.spent - envelope.committed);
    const utilizationPercent =
      envelope.limit > 0 ? Math.min(100, (envelope.spent / envelope.limit) * 100) : 0;

    return {
      category: envelope.category,
      limit: envelope.limit,
      spent: envelope.spent,
      committed: envelope.committed,
      available,
      utilizationPercent,
      period: envelope.period,
      resetAt: envelope.resetAt,
    };
  }

  /**
   * Returns utilisation snapshots for all registered categories.
   */
  listUtilizations(): readonly BudgetUtilization[] {
    return Array.from(this.#envelopes.keys())
      .map((category) => this.getUtilization(category))
      .filter((u): u is BudgetUtilization => u !== undefined);
  }
}
